Enable test in parallel with Helix (and non helix and helix test also in parallel)

wli3 opened the pull request:
Review via commits helps

Issue for improvements https://github.com/dotnet/sdk/issues/14530

Commit:
re-enable helix

Commit:
Fix bad merge

Commit:
Fix test assertion by normalize path

Commit:
Make websdk tests be able to run on the host

Make websdk tests to have the same test project as the rest of
sdk test proejcts

Commit:
Passing parameters and files from the host machine

Commit:
Test target to help local repro

Commit:
Use relative path for SDK resolver xml

Commit:
Remove not used import

Commit:
Run helix and non helix in parallel

dsplaisted commented on src/Tests/Microsoft.NET.TestFramework/TestCommandLine.cs:
```diff
@@ -29,10 +29,10 @@ public class TestCommandLine
-
         public string TestExecutionDirectory { get; set; }
 
+        public string MsbuildAdditionalSdkRsolverFolder { get; set; }
```
Typo (here and elsewhere, including command line help)

```suggestion
        public string MsbuildAdditionalSdkResolverFolder { get; set; }
```

dsplaisted commented on src/Tests/Microsoft.NET.TestFramework/ToolsetInfo.cs:
```diff
@@ -277,8 +282,19 @@ public static ToolsetInfo Create(string repoRoot, string repoArtifactsDir, strin
+                }
+                else
+                {
+                    throw new ApplicationException("Microsoft.DotNet.MSBuildSdkResolver path is not provided, set msbuildAdditionalSdkRsolverFolder on test commandline or set repoRoot");
```
Nit: I typically use InvalidOperationException for this kind of thing.  I think ApplicationException is sort of soft-deprecated.

dsplaisted commented on src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/Microsoft.DotNet.MSBuildSdkResolver.csproj:
```diff
@@ -33,13 +33,18 @@
 
   <Target Name="WriteFullFrameworkResolverManifest" AfterTargets="AfterBuild" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
+    <PropertyGroup>
+      <_ResolverManifestDir>$(OutputPath)\SdkResolvers\$(AssemblyName)\</_ResolverManifestDir>
```
I'd suggest adding a comment here explaining what's going on.  It sounds like this is getting the relative path to put in the XML file so it still works in Helix when the absolute paths are different.

dsplaisted reviewed (commented):
I haven't made it through all the commits yet, but I figured I'd give you the feedback I have so far.

Have you tested that test failures are reported properly (both for tests on helix and non-helix tests).  Can the failures be viewed in the Azure DevOps UI in both cases?

How possible is it to locally test in Helix mode?  Is there any write-up on how to do that?

wli3 commented on src/Tests/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs:
```diff
@@ -72,7 +79,14 @@ public override bool Execute()
+                .Where(l => !string.IsNullOrWhiteSpace(l));
+
+            var xUnitProjectsCanRunInHelix
+                = XUnitProjects.Where(p => !testExclusion.Any(p.ItemSpec.Contains));
```
There is no easy reliable way to do that other than regex in both powershell and bash. It is better to keep the logic in sync, since overlap or missing test projects are worse than accidentally run on host.

Commit:
Use DependsOnTargets instead

wli3 commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToReferenceAProject.cs:
```diff
@@ -203,16 +203,17 @@ public void It_disables_copying_conflicting_transitive_content(bool copyConflict
             getValuesCommand.Execute().Should().Pass();
 
-            var valuesResult = getValuesCommand.GetValuesWithMetadata().Select(pair => pair.value);
+            var valuesResult = getValuesCommand.GetValuesWithMetadata().Select(pair => Path.GetFullPath(pair.value));
```
in the test, the path are constructed in 2 different ways. "pair.value" will not "expand". For example you have "C:\Users\\**ADMINI~1**\\AppData\Local\Temp\dotnetSdkTests\p5eoajv5.iq5\It_builds_a_s---34401EFE\"   vs "C:\Users\\administrator\....."

Commit:
Correct typo

Commit:
Use InvalidOperationException

Commit:
Add comment for the relative path

wli3 commented:
I addressed the comment so far.

> Have you tested that test failures are reported properly (both for tests on helix and non-helix tests). Can the failures be viewed in the Azure DevOps UI in both cases?

Yes. Since I've been using DevOps UI to investigate what is failing. All the test report will still end up in the same folder as before. And it will be collected by Azure Devops.

> How possible is it to locally test in Helix mode? Is there any write-up on how to do that?

I added a hook to make it much easier to repro helix issue. I'm working on the write up.

Commit:
Add doc to repro helix failure

wli3 commented:
@dsplaisted the latest commit added how to repro helix failure locally https://github.com/dotnet/sdk/pull/14532/commits/dcd9a9d5f5d6f6e8af83d13a15afb824f5ad7d0c

Commit:
Use Windows.Server.Amd64.VS2019.Pre.Open

wli3 commented:
@dsplaisted the new queue is online now. Helix and the host should have the same configuration(same VS version). Xcopy msbuild changes are reverted

dsplaisted commented on eng/runHelixAndNonHelixInParallel.ps1:
```diff
@@ -0,0 +1,40 @@
+        [switch] $test
+    )
+
+    if (-not$test)
```
Should there be a space here?

```suggestion
    if (-not $test)
```

wli3 replied:
My automatic formatter did that. I guess it is a convention somewhere

dsplaisted commented on eng/runTestsCannotRunOnHelix.ps1:
```diff
@@ -1,3 +1,6 @@
+#hack to make sure build of helix run is done, so no file lock
+Start-Sleep -Seconds 90
```
I'd love to see us eventually get rid of this hack.

wli3 replied:
Agree, we should eventually move all tests to helix. And we can remove that part completely.

dsplaisted commented on documentation/project-docs/repro-helix-failure.md:
```diff
@@ -0,0 +1,65 @@
+
+## General
+
+Most of the test failure on helix should be able to repro locally. Try follow [developer-guide](developer-guide.md) _Run tests from the command line_ session first.
```
```suggestion
Most test failures on helix should repro locally. Try following the [developer-guide](developer-guide.md) _Run tests from the command line_ section first.
```

dsplaisted commented on documentation/project-docs/repro-helix-failure.md:
```diff
@@ -0,0 +1,65 @@
+
+## Repro test as tool layout
+
+Run test on Helix reuse "test as tool" infrastructure. All tests projects are published with xunit command line. The SDK under test, the published test project and other dependencies are copied to a helix machine. This step changed the repo structure. Most of the helix-only failures are caused by tests depending on the repo structure or having dependencies that are not carried by the test project ...
```
```suggestion
Running tests on Helix reuses the "test as tool" infrastructure. All tests projects are published  as tools that accept xunit command line arguments. The SDK under test, the published test project, and other dependencies are copied to a helix machine. This step changes the repo structure. Most of the helix-only failures are caused by tests depending on the repo structure or having dependencies that are not carried by the test project publishing step. The following steps help you to repro the test layout that is used on the helix machine. And you can run the individual test to discover the cause.
```

dsplaisted commented on documentation/project-docs/repro-helix-failure.md:
```diff
@@ -0,0 +1,65 @@
+
+## Folders under localHelixTestLayout
+
+Due to Helix long path problem. The folders under localHelixTestLayout all have one letter name.
```
Can we enable long paths on Helix machines?  We should try to avoid having to workaround long path issues, as we're likely to eventually go over the max path and have hard-to-diagnose issues.

wli3 replied:
I had a long talk about the long path issue. Last year for at least a month. (We first found it 2 years ago for the perf tests and started to ask for it) it is not easy, due to the way it is configured. Although, we get the path on helix trimmed, it used to have 2 guid in front.

dsplaisted commented on documentation/project-docs/repro-helix-failure.md:
```diff
@@ -0,0 +1,65 @@
+
+## Other possible cause of Helix-only failure
+
+- Helix machine are not the same configuration as the host machine. Ideally the dependency should be declared by the SDK repo itself. At the time the document is written. The build queue and the helix queue should have the same configuration. If they are diverged, we should contact arcade team.
```
What do you mean by configuration here?  Build vs debug?

wli3 replied:
Used "pre-installed dependencies" instead. It is mainly about VS version

Commit:
Address code review

wli3 commented:
@dsplaisted thank you for the review. I addressed them. I'll merge once the CI is done

dsplaisted commented:
> @dsplaisted thank you for the review. I addressed them. I'll merge once the CI is done

And the CI should complete more quickly than usual because of these changes! 🎉🎉🎉
